docs/gcloud-auth.rst

```diff
@@ -0,0 +1,264 @@
+==============
+
+**If you're running in Compute Engine or App Engine**,
+authentication should "just work"
+as credentials can be loaded automatically
+and therefore there is no extra code required to authenticate.
+
+**If you're developing locally**,
+the easiest way to authenticate is using the Cloud SDK.
+
+**If you're running your application elsewhere**,
+you should download the service account JSON keyfile
+and point to it using an environment variable.
+
+-----------------------------------------------
+In what order does gcloud discover credentials?
+-----------------------------------------------
+
+``gcloud`` will look in a bunch of different places
+for authentication credentials, in the following order:
+
+#. Credentials you explicitly set in code
+#. Credentials from an environment variable (``GOOGLE_APPLICATION_CREDENTIALS``)
+#. Credentials provided by the Cloud SDK (``gcloud auth login``)
+#. Credentials provided when running in Google App Engine
+#. Credentials provided when running in Google Compute Engine
+
+This means that you can set your credentials a bunch of different ways.
+We recommend using some form of implicitly discovered credentials
+(ie, set in an environment variable or auto-discovered in GCE/GAE)
+so that your code can be written the same and
+simply run in the environment of your choice.
+
+=========================================
+Authenticating locally with the Cloud SDK
+=========================================
+
+This is the easiest way to authenticate while you're developing locally.
+
+#. Download and install the Cloud SDK (You can get the Cloud SDK at https://cloud.google.com/sdk)
+#. Authenticate using OAuth2 (``gcloud auth login``)
+#. Run your code (ie, ``python myscript.py``)
+
+
+After you do this,
+your script will look for the authentication token
+that you got when authenticating,
+and you won't need to think about authentication in your code at all.
+
+======================================
+Authenticating with a specific keyfile
+======================================
+
+----------------------------------------------
+Setting the keyfile in an environment variable
+----------------------------------------------
+
+If you have a specific JSON keyfile downloaded that you'd like to use,
+you can simply set the path to this in
+the ``GOOGLE_APPLICATION_CREDENTIALS`` environment variable.
+
+This means your code doesn't have to change at all,
+and can run with different credentials depending on the environment.
+
+Here's what this looks like:
+
+.. code-block:: bash
+
+   $ export GOOGLE_APPLICATION_CREDENTIALS="/path/to/key.json"
+   $ python
+   >>> from gcloud import pubsub
+   >>> pubsub.Topic('topic_name').create()
+
+--------------------------------------
+Setting the keyfile explicitly in code
+--------------------------------------
+
+If you really want to set a specific keyfile in code
+(maybe for a code snippet to send to a friend?)
+you can do this, but it's really not recommended.
```

Comment:
I tend to agree with you that it can be confusing to have lots of magic (hence why I really really want these options described here).

What would the ideal auth story look like if you guys could pick anything you want? (code) @dhermes @tseaver